targets/psa_switch/externs/psa_checksum.cpp

```diff
@@ -0,0 +1,69 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_checksum.h"
+
+namespace bm {
+
+namespace psa {
+
+void
+PSA_Checksum::init() {
+  this->clear();
+}
+
+void
+PSA_Checksum::get(Field& dst) const {
+  dst.set(internal.get<uint64_t>());
+}
+
+void
+PSA_Checksum::get_verify(Field& dst, Field& equOp) const {
+  if (equOp.get<uint64_t>()!=internal.get<uint64_t>()) {
+    dst.set(false);
+  } else {
+    dst.set(true);
+  }
+}
+
+void
+PSA_Checksum::clear() {
+  internal.set(0);
+}
+
+void
+PSA_Checksum::update(const NamedCalculation& calculation) {
+  const uint64_t cksum = calculation.output(get_packet());
+  this->internal.set(cksum);
+}
```

Comment:
@DusanKrdzic I was looking at this one more time and I am not sure this implements the actual PSA specification. `update` is a method you can call multiple times for a given checksum calculation, in order to support incremental computation. Unless I am mistaken, it seems that your current implementation assumes a single call to `update`, as every call overrides the `internal` value. Am I missing something?

Reply:
This is the `basic checksum` implementation with PSA extern `extern Checksum`. I couldn't find anywhere in the `PSA spec` that the `basic checksum` uses incremental computation to compute checksum. I think it use basic computation as the name says.
There is a special type of checksum in PSA `extern InternetChecksum` which uses incremental computation. It has special method `add` for that kind of calculation.

Reply:
`Basic checksum -  Checksum`
![basic_checksum](https://user-images.githubusercontent.com/77269536/122342099-0de19c00-cef9-11eb-965d-21b55b1e8f2b.png)

`Incremental checksum -  InternetChecksum`
![incremental](https://user-images.githubusercontent.com/77269536/122342260-46817580-cef9-11eb-982c-a3569edeac3e.png)

Reply:
let's double-check with @jafingerhut.

@jafingerhut I want to make sure what the expected behavior is for the `update` extern method:
1. overwrite the checksum value, using only the data provided in the `update` call
2. append the data to an internal buffer (preserving the data provided in previous `update` calls), the checksum is computed on the entire buffer when `get` is called

My understanding of the spec is that the second one is the expected behavior:
```
  /// Add data to checksum
  void update<T>(in T data);

  /// Get checksum for data added (and not removed) since last clear
  W    get();
```

The "and not removed" is a bit confusing, since this extern doesn't support removing data.

Reply:
My understanding of the spec is that the first is expecting behavior because it is the basic checksum. If basic checksum also uses incremental computation why is extern InternetChecksum needed? I understood that this update is some kind of computation that is used in simple switch which uses basic checksum computation. In PSA spec it says "PSA also provides an incremental checksum" after defining the basic checksum. So there must be a difference in computation.

Reply:
Maybe incremental is not the right term for what the basic checksum does. It doesn't have a `remove` method. With the `InternetChecksum`, if the value of a field changes, you can update the checksum by removing the old value and adding the new value. That's not possible with `Checksum`, but I still believe the intent is to support multiple `update` calls. Anyway, I will wait for @jafingerhut to chime in.

Reply:
InternetChecksum is explicitly incremental, not only for adding more data, but also for removing data, so that it can even be used for implementing things like updating TCP and UDP header checksums in certain situations.  There are both add and subtract methods for InternetChecksum for this very reason.

Antonin, I agree that the "and not removed" part is confusing, and to me it is not terribly explicit in the latest PSA spec whether behavior 1 or 2 is being specified.  It has been about 3 years since there were discussions on this topic, and I do not now recall any discussions that would favor interpretation 1 over 2, or vice versa.  Let me check on the details of at least one PSA implementation of Checksum extern in this regard and report back.

Reply:
If it is not terribly explicit why it shouldn't be the first interpretation:
`1. overwrite the checksum value, using only the data provided in the update call`
because it has already been implemented with this pull-request. 
I think that the name `basic` defines that it uses the first interpretation. If someone wants to use incremental computation then there is special extern type `InternetChecksum` for that kind of computation.

Reply:
@DusanKrdzic "because this PR does it this way" is not a good enough reason. Let's wait to hear back from Andy on the status of current implementations.

@jafingerhut we should probably update the spec and the method comments for `update` and `get`.

Reply:
@antoninbas Agreed that the spec should be clarified here, regardless of the answer.

Reply:
To improve the computation of large chunk of data such as in TCP, RFC 1071 (https://datatracker.ietf.org/doc/html/rfc1071) introduced `InternetChecksum` with incremental computing with `1’s complement` algorithm which is especially used for incremental computation. This technique is covered in RFC 1071 and further polished over two other RFCs: RFC 1141 (https://datatracker.ietf.org/doc/html/rfc1141) and RFC 1624 (https://datatracker.ietf.org/doc/html/rfc1624).
For the same reason `InternetChecksum` is also introduced in PSA as incremental checksum with `1's complement` to perform incremental computation. `InternetChecksum` uses `only` `1's complement`  algorithm because of explicit incremental computation (like in RFC).
In PSA, `Basic checksum` can use all other hash algorithms which are not explicitly used for incremental computation.
This is one of the reasons why I think that there is a difference in computation between `basic` and `internet` checksum.
If `basic` checksum also uses incremental computation why is it needed? Then `InternetChecksum` would do the same even much more (because of `substract` and other methods that `basic` checksum doesn't have). In this case `basic` checksum would be a kind of subset of `InternetChecksum`. This would make `basic` checksum redundant.